lms/djangoapps/ccx/modulestore.py

```diff
@@ -0,0 +1,298 @@
+    if hasattr(val, 'location'):
+        val.location = restore_ccx(val.location, ccx_id)
+    if hasattr(val, 'children'):
+        val.children = restore_ccx_collection(val.children, ccx_id)
+    return val
+
+
+def restore_ccx_collection(field_value, ccx_id=None):
+    """restore references to a CCX to collections of incoming values
+
+    returns the original collection with all values converted to a ccx-aware
+    state, using the provided ccx_id
+    """
+    if ccx_id is None:
+        return field_value
+    if isinstance(field_value, list):
+        field_value = [restore_ccx(fv, ccx_id) for fv in field_value]
+    elif isinstance(field_value, dict):
+        for key, val in field_value.iteritems():
+            field_value[key] = restore_ccx(val, ccx_id)
+    else:
+        field_value = restore_ccx(field_value, ccx_id)
+    return field_value
+
+
+@contextmanager
+def remove_ccx(to_strip):
+    """A context manager for wrapping modulestore api methods.
+
+    yields a stripped value and a function suitable for restoring it
+    """
+    stripped, ccx = strip_ccx(to_strip)
+    yield stripped, partial(restore_ccx_collection, ccx_id=ccx)
+
+
+class CCXModulestoreWrapper(object):
+    """This class wraps a modulestore
+
+    The purpose is to remove ccx-specific identifiers during lookup and restore
+    it after retrieval so that data can be stored local to a course, but
+    referenced in app context as ccx-specific
+    """
+
+    def __init__(self, modulestore):
+        """wrap the provided modulestore"""
+        self.__dict__['_modulestore'] = modulestore
+
+    def __getattr__(self, name):
+        """look up missing attributes on the wrapped modulestore"""
+        return getattr(self._modulestore, name)
+
+    def __setattr__(self, name, value):
+        """set attributes only on the wrapped modulestore"""
+        setattr(self._modulestore, name, value)
+
+    def __delattr__(self, name):
+        """delete attributes only on the wrapped modulestore"""
+        delattr(self._modulestore, name)
+
+    def _clean_locator_for_mapping(self, locator):
+        """See the docs for xmodule.modulestore.mixed.MixedModuleStore"""
+        with remove_ccx(locator) as (locator, restore):
+            # pylint: disable=protected-access
+            return restore(
+                self._modulestore._clean_locator_for_mapping(locator)
+            )
+
+    def _get_modulestore_for_courselike(self, locator=None):
+        """See the docs for xmodule.modulestore.mixed.MixedModuleStore"""
+        if locator is not None:
+            locator, _ = strip_ccx(locator)
+        # pylint: disable=protected-access
+        return self._modulestore._get_modulestore_for_courselike(locator)
+
+    def fill_in_run(self, course_key):
+        """See the docs for xmodule.modulestore.mixed.MixedModuleStore"""
+        with remove_ccx(course_key) as (course_key, restore):
+            return restore(self._modulestore.fill_in_run(course_key))
```

Comment:
@cpennington @cewing The tests are old but they are failing because we introduced a new bookmarks service in module render and that service is calling `fill_in_run` method & currently they are fail in case of `XMLModuleStore` objects.